spatial_cells/block_adjust_gpu.cpp

```diff
@@ -754,4 +614,266 @@ void adjust_velocity_blocks_in_cells(
+#pragma omp for schedule(dynamic)
+      for (size_t i=0; i<nCells; ++i) {
+         SpatialCell* SC = mpiGrid[cellsToAdjust[i]];
+         if (SC->sysBoundaryFlag == sysboundarytype::DO_NOT_COMPUTE) {
+            continue;
+         }
+         // Grow mesh if necessary and on-device resize did not work??
+         const vmesh::LocalID nBlocksBeforeAdjust = host_nBefore[i+cellOffset];
+         const vmesh::LocalID nBlocksAfterAdjust  = host_nAfter[i+cellOffset];
+         const vmesh::LocalID nBlocksToChange     = host_nBlocksToChange[i+cellOffset];
+         const vmesh::LocalID resizeDevSuccess    = host_resizeSuccess[i+cellOffset];
+         thread_largestBlocksToChange = std::max(thread_largestBlocksToChange, nBlocksToChange);
+         // This is gathered for mass loss correction: for each cell, we want the smaller of either blocks before or after. Then,
+         // we want to gather the largest of those values.
+         const vmesh::LocalID lowBlocks = std::min(nBlocksBeforeAdjust, nBlocksAfterAdjust);
+         thread_largestBlocksBeforeOrAfter = std::max(thread_largestBlocksBeforeOrAfter, lowBlocks);
+         if ( (nBlocksAfterAdjust > nBlocksBeforeAdjust) && (resizeDevSuccess == 0)) {
+            //GPUTODO is _FACTOR enough instead of _PADDING?
+            SC->get_velocity_mesh(popID)->setNewCapacity(nBlocksAfterAdjust*BLOCK_ALLOCATION_PADDING);
+            SC->get_velocity_mesh(popID)->setNewSize(nBlocksAfterAdjust);
+            SC->get_velocity_blocks(popID)->setNewCapacity(nBlocksAfterAdjust*BLOCK_ALLOCATION_PADDING);
+            SC->get_velocity_blocks(popID)->setNewSize(nBlocksAfterAdjust);
+            SC->dev_upload_population(popID);
+         }
+         // Update cached sizes
+         SC->get_velocity_mesh(popID)->setNewCachedSize(nBlocksAfterAdjust);
+         SC->get_velocity_blocks(popID)->setNewCachedSize(nBlocksAfterAdjust);
+      } // end cell loop
+#pragma omp critical
+      {
+         largestBlocksToChange = std::max(thread_largestBlocksToChange, largestBlocksToChange);
+         largestBlocksBeforeOrAfter = std::max(largestBlocksBeforeOrAfter, thread_largestBlocksBeforeOrAfter);
+      }
+   } // end parallel region
+   hostResizeTimer.stop();
+   // Writing directly into pass-by-reference variables from within OMP parallel region caused issues
+   out_largestBlocksToChange = largestBlocksToChange;
+   out_largestBlocksBeforeOrAfter = largestBlocksBeforeOrAfter;
+
+   // Do we actually have any changes to perform?
+   if (largestBlocksToChange > 0) {
+      phiprof::Timer addRemoveKernelTimer {"GPU batch add and remove blocks kernel"};
+      // Third argument specifies the number of bytes in *shared memory* that is
+      // dynamically allocated per block for this call in addition to the statically allocated memory.
+      dim3 grid_addremove(largestBlocksToChange,nCells,1);
+      batch_update_velocity_blocks_kernel<<<grid_addremove, WID3, 0, baseStream>>> (
```

Comment:
In a similar vein as above, for the sake of future optimisation hunters, any comments with wisdom as to the choices of grids and blocks would be very welcome. Now here it is kind of self-explanatory I guess, but nevertheless.